### ISSUE TITLE
onZoom, onPan events to early

### ISSUE THREAD

**Issue by dronus** [ariutta/svg-pan-zoom#121]
As the SVG's transformation matrix is updated deferred (with timeouts or animation frame handlers I guess), the SVG does not represent the current state if onZoom or onPan fires. 

For example, properties of the SVG content like bounding boxes depending on the view do not reflect the new view. 

This makes it hard to synchronize other elements to SVG elements (like a HTML Popup tied to an SVG element) or hide / show SVG elements depending on the viewport.

The events should fire after the new transform is applied.

**Comment by bumbu** [comment]
This is a desired behavior as in this case you don't have to worry about limiting the number of updates/second that are done. Also it _seems to_ improve performance on slower machines.

The delta time between onZoom/onPan events and actual SVG update should be on average less than 40ms (depends on your browser, available resources and timeout settings). This amount of time is almost invisible for a human eye, so if you sync a popup with SVG state - it is very improbable that user will notice any lag.

There are cases where performance may be bad (like in [this question](http://stackoverflow.com/q/30161282/1194327)) but in most cases it is because of SVG rendering. So even if onZoom/onPan event would be fired after update of transform matrix values user would see a few seconds lag because of SVG slow render. A solution would be to fire such events only after SVG finished rendering, but I don't know if this is possible.

It would be great if you could show an example where this behavior (firing onZoom/onPan) before applying changes to transform matrix causes a buggy/laggy behavior.

One solution I do see is to add something like onZoomApply/onPanApply or onZoomRender/onPanRender that would fire after actual update of transform matrix.

**Comment by dronus** [comment]
I have no problem with deffering the matrix update. Multiple transform events may be coerces too. However, the handlers should not fire before the update was made.

My use case was to  calculate the field of view, and add svg objects on the fly depending on the current view for example. In this case, the new objects lag behind, eg. there are areas at the screen borders where the objects where not created. 

A dirty hack in this case was to replace the viewports' setAttributeNS setter to capture the matrix if set. Not very nice though...

Also other DOM objects (eg, HTML elements) may be placed to match some of the svg's elements, and those would lag behind too, if depending on the transform matrix.

**Comment by Rush** [comment]
I too encountered this issue. Solution was to add my own `onUpdate` callback. I didn't do a pull request yet since I see pull requests are not welcome here.

```js
ShadowViewport.prototype.updateCTM = function() {
  // Updates SVG element
  SvgUtils.setCTM(this.viewport, this.getCTM(), this.defs);
  if(this.options.onUpdateCTM) this.options.onUpdateCTM();
```

```js
  , onPan: function(point) {
      if (that.viewport && that.options.onPan) {return that.options.onPan(point)}
    }
  , onUpdateCTM: function(ctm) {
      if (that.viewport && that.options.onUpdate) {return that.options.onUpdate(ctm)}
    }
```

**Comment by bumbu** [comment]
Hi @Rush , 

I'm sorry to hear that you have an impression that pull-requests are not welcome.
They're are very welcome, and we have [instructions](https://github.com/ariutta/svg-pan-zoom#contributionspull-requests) on how to do that. Many libraries don't bother doing that. 

It's just often pull-requests are of poor quality, or try to bring breaking changes. Many pull-requests include files that shouldn't be included, that's why we just cherry-pick essential changes, commit them and close original pull requests. Most pull-requests find their way into the codebase (and most of them were cherry-picked). 

Adding an `onUpdatedCTM` callback seems a very good solution to this problem as it will not break any  existing code while allowing covering use-cases mentioned in this thread. I'd opt for `onUpdatedCTM` instead of `onUpdateCTM` in order to make it a bit more clear that that's a post-effect callback. 

Also would be good to have it called after the lock is released (so that to avoid potential problems if someone is going to call pan or zoom from the inside of the `onUpdated`):
```
ShadowViewport.prototype.updateCTM = function() {
  // Updates SVG element
  SvgUtils.setCTM(this.viewport, this.getCTM(), this.defs)

  // Free the lock
  this.pendingUpdate = false
 
  // Notify about the update
  if(this.options.onUpdatedCTM) this.options.onUpdatedCTM();
}
```

Please do a pull-request. But please follow [the instructions](https://github.com/ariutta/svg-pan-zoom#contributionspull-requests) and don't commit `dist` files. Also we'd appreciate if you'll add some tests for that new behavior. It will also make much easier for us to review and merge the code.
